File: src/main/java/edu/metrostate/trackingsystem/application/services/DealerService.java
package edu.metrostate.trackingsystem.application.services;

import edu.metrostate.trackingsystem.domain.models.Dealer;
import edu.metrostate.trackingsystem.domain.repositories.DealerRepository;
import edu.metrostate.trackingsystem.infrastructure.database.Result;
import javafx.collections.ObservableList;

import java.util.List;

public class DealerService {

    private final DealerRepository repository;

    public DealerService(DealerRepository repository) {
        this.repository = repository;
    }

    public List<String> getDealershipIDs() {
        return repository.getDealershipIDs();
    }

    public void toggleAcquisition(String dealershipId) {
        repository.toggleAcquisition(dealershipId);
    }

    public ObservableList<Dealer> getDealers() {
        return repository.getDealers();
    }

    public Result<Boolean> deleteDealer(String dealershipId) {
        return repository.deleteDealer(dealershipId);
    }

    public boolean updateDealer(String dealershipId, String name) { return repository.updateDealer(dealershipId, name); }



}

File: src/main/kotlin/edu/metrostate/dealership/application/services/DealerService.kt
//This class is used for transferring data through application layers from the database
package edu.metrostate.dealership.application.services

import edu.metrostate.dealership.domain.models.Dealer
import edu.metrostate.dealership.domain.repositories.DealerRepository
import edu.metrostate.dealership.infrastructure.database.Result
import javafx.collections.ObservableList

class DealerService(private val repository: DealerRepository) {
    val dealershipIDs: List<String>
        get() = repository.dealershipIDs

    fun toggleAcquisition(dealershipId: String?) {
        repository.toggleAcquisition(dealershipId)
    }

    fun transferInventory() {}
    val dealers: ObservableList<Dealer>
        get() = repository.dealers

    fun deleteDealer(dealershipId: String?): Result<Boolean>? {
        return repository.deleteDealer(dealershipId)
    }

    fun updateDealer(dealershipId: String?, name: String?): Unit? {
        return repository.updateDealer(dealershipId, name)
    }


    /**
     * Adds a new dealer to the system.
     * Validates that the dealer ID doesn't already exist.
     *
     * @param dealer The dealer object to add
     * @return The added dealer
     * @throws IllegalArgumentException If a dealer with the same ID already exists
     */
    fun addDealer(dealer: Dealer): Dealer {
        if (dealershipIDs.contains(dealer.dealershipId)) {
            throw IllegalArgumentException("A dealer with ID ${dealer.dealershipId} already exists.")
        }

        repository.dealers.add(dealer)
        return dealer
    }
}